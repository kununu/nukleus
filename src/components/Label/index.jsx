import React from 'react';
import PropTypes from 'prop-types';

import {
  controlLabel,
  srOnly,
} from '../index.scss';

export default function Label ({
  id, value, labelHidden, classNames, isTitle,
}) {
  const hidden = labelHidden ? srOnly : '';
  const allClassNames = `${classNames} ${controlLabel} ${hidden}`;

  /**
   * For checkbox groups or radio button groups you may want
   * to have a title that looks like a label
   *
   * @return {ReactElement} [Returns a div with the label styles]
  */
  if (isTitle) {
    return <div className={allClassNames}>{value}</div>;
  }

  /**
   * generates a label or div representing the label of a group of fields
   *
   * @return {ReactElement} [Either returns a label or div or a react element]
  */
  if (typeof value === 'string') {
    return (
      <label
        htmlFor={id}
        className={allClassNames}
      >
        {value}
      </label>
    );
  }

  // We don't simply put a more complex element inside a label to prevent a
  // clickable element like a link or button inside a label
  // However to also add the labelContainer class, we need to return a cloned
  // element and not just the label - element itself
  return React.cloneElement(
    value,
    {
      ...value.props,
      className: allClassNames,
    },
  );
}

Label.propTypes = {
  classNames: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
  ]),
  id: PropTypes.string,
  isTitle: PropTypes.bool,
  labelHidden: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
  ]).isRequired,
};

Label.defaultProps = {
  classNames: '',
  id: '',
  isTitle: false,
  labelHidden: false,
};
