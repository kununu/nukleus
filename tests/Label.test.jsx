import React from 'react';
import renderer from 'react-test-renderer';
import Label from 'Label'; // eslint-disable-line import/no-unresolved

test('Renders a label without crashing', () => {
  const component = renderer.create(<Label
    id="2"
    value="Test label"
    classNames="test-class"
  />);

  expect(component.toJSON()).toMatchSnapshot();
});

test('Renders a title instead of a label', () => {
  const component = renderer.create(<Label
    value="Test label"
    classNames="test-class"
    isTitle
  />);

  expect(component.toJSON()).toMatchSnapshot();
});

test('Renders a hidden label', () => {
  const component = renderer.create(<Label
    id="2"
    value="Test label"
    labelHidden
    classNames="test-class"
    isTitle
  />);

  expect(component.toJSON()).toMatchSnapshot();
});
