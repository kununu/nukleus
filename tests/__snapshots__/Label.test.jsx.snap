exports[`test Renders a hidden label 1`] = `
<div
  className="test-class controlLabel srOnly">
  Test label
</div>
`;

exports[`test Renders a label without crashing 1`] = `
<label
  className="test-class controlLabel "
  htmlFor="2">
  Test label
</label>
`;

exports[`test Renders a title instead of a label 1`] = `
<div
  className="test-class controlLabel ">
  Test label
</div>
`;
